dbt_coverage/__init__.py

```diff
@@ -15,82 +15,129 @@
-        return documented, total
+    def coverage(self, cov_type: CoverageType):
+        coverages = [col.coverage(cov_type) for col in self.columns.values()]
+        covered = set(f'{self.name}.{col}' for covered, _ in coverages for col in covered)
+        total = set(f'{self.name}.{col}' for _, total in coverages for col in total)
+        return covered, total
 
 
 @dataclass
-class Catalog:
+class Catalog(EntityWithCoverage):
     """Dataclass containing the information about a database catalog, its tables and columns."""
 
     tables: Dict[str, Table]
 
     @staticmethod
-    def from_catalog_nodes(nodes):
-        return Catalog({f"{node['metadata']['name']}": Table.from_catalog_node(node)
-                        for node in nodes})
-
-    @staticmethod
-    def from_manifest_nodes(nodes):
-        return Catalog({f"{node['name']}": Table.from_manifest_node(node)
+    def from_nodes(nodes):
+        return Catalog({f"{node['metadata']['name']}": Table.from_node(node)
                         for node in nodes})
 
     def get_table(self, table_name):
         return self.tables.get(table_name)
 
-    def get_coverage(self):
-        coverages = [(table, table.get_coverage()) for table in self.tables.values()]
-        documented = set(f'{table.name}.{col}'
-                         for table, (documented, _) in coverages
-                         for col in documented)
-        total = set(f'{table.name}.{col}'
-                    for table, (_, total) in coverages
-                    for col in total)
-        return documented, total
+    def coverage(self, cov_type: CoverageType):
+        coverages = [table.coverage(cov_type) for table in self.tables.values()]
+        covered = set(col for covered, _ in coverages for col in covered)
+        total = set(col for _, total in coverages for col in total)
+        return covered, total
+
+
+@dataclass
+class Manifest:
+    sources: Dict[str, Dict[str, Dict]]
+    models: Dict[str, Dict[str, Dict]]
+    tests: Dict[str, Dict[str, List[Dict]]]
+
+    @classmethod
+    def from_nodes(cls, manifest_nodes: Dict[str: Dict]) -> Manifest:
+        id_to_table_name = {table_id: table['name'] for table_id, table in manifest_nodes.items()
+                            if table['resource_type'] in ['source', 'model']}
+        sources = {table['name']: table['columns'] for table in manifest_nodes.values()
+                   if table['resource_type'] == 'source'}
+        models = {table['name']: table['columns'] for table in manifest_nodes.values()
+                  if table['resource_type'] == 'model'}
+        tests = {}
+        for node in manifest_nodes.values():
+            if node['resource_type'] != 'test' or 'schema' not in node['tags']:
+                continue
+
+            depends_on = node['depends_on']['nodes']
+            if node['test_metadata']['name'] == 'relationships':
+                table_id = depends_on[len(depends_on) - 1]
```

Comment:
@sweco this seems rather fragile, right?

Reply:
Yes, but it's copied from `dbt-docs`: https://github.com/dbt-labs/dbt-docs/blob/da84e95ab9febde3127ed28ff698516b2a610532/src/app/services/project_service.js#L201